ykrt/src/compile/jitc_yk/codegen/x86_64.rs

```diff
@@ -0,0 +1,252 @@
+//! The X86_64 JIT Code Generator.
+
+use super::super::jit_ir::{self, InstrIdx, Operand};
+#[cfg(any(debug_assertions, test))]
+use super::AsmPrinter;
+use super::{CodeGen, LocalAlloc};
+use dynasmrt::{dynasm, x64::Rq, AssemblyOffset, DynasmApi, Register};
+#[cfg(any(debug_assertions, test))]
+use iced_x86;
+
+/// Argument registers as defined by the X86_64 SysV ABI.
+static ARG_REGS: [Rq; 6] = [Rq::RDI, Rq::RSI, Rq::RDX, Rq::RCX, Rq::R8, Rq::R9];
+
+/// The size of a 64-bit register in bytes.
+static REG64_SIZE: usize = 8;
+
+/// Work registers, i.e. the registers we use temproarily (where possible) for operands to, and
+/// results of, intermediate computations.
+///
+/// We choose callee-save registers so that we don't have to worry about storing/restoring them
+/// when we do a function call to external code.
+static WR0: Rq = Rq::R12;
+static WR1: Rq = Rq::R13;
+static WR2: Rq = Rq::R14;
+
+/// The X86_64 SysV ABI requires a 16-byte aligned stack prior to any call.
+const SYSV_CALL_STACK_ALIGN: usize = 16;
+
+impl CodeGen<'_> {
```

Comment:
Ah, I see that you're effectively extending `CodeGen` here. The problem with this is that it means you can only ever have one code gen compiled in. That means we can't ever have e.g. a slow & fast code gen that we swap between at run-time. It might make testing more complex too?

Probably the machine-dependent part should `impl` a `trait` of some kind.

Reply:
Yep, as discussed above and in comments in the code. One for a follow up PR.

Reply:
I think we should improve this in this PR, because it's quite a fundamental decision.

Reply:
OK.

Reply:
Fixed in https://github.com/ykjit/yk/pull/971/commits/4c25a723383a8c577fad6d1f339b8dd37bd0c88e